epuck_ros2_camera/src/pipuck_mmal.c

```diff
@@ -0,0 +1,187 @@
+  assert(status == MMAL_SUCCESS);
+
+  // Create a buffer pool
+  context->queue = mmal_queue_create();
+  pipuck_mmal->internal.pool_in = mmal_port_pool_create(context->encoder->input[0], context->encoder->input[0]->buffer_num,
+                                                        context->encoder->input[0]->buffer_size);
+  assert(context->pool_in != 0);
+  context->pool_out = mmal_port_pool_create(context->encoder->output[0], context->encoder->output[0]->buffer_num,
+                                            context->encoder->output[0]->buffer_size);
+  assert(context->pool_out != 0);
+
+  // Start component
+  status = mmal_component_enable(context->encoder);
+  assert(status == MMAL_SUCCESS);
+}
+
+void pipuck_mmal_deinit(pipuck_mmal_t *pipuck_mmal) {
+  mmal_port_disable(pipuck_mmal->internal.encoder->input[0]);
+  mmal_port_disable(pipuck_mmal->internal.encoder->output[0]);
+  mmal_component_disable(pipuck_mmal->internal.encoder);
+  mmal_queue_destroy(pipuck_mmal->internal.queue);
+  mmal_port_pool_destroy(pipuck_mmal->internal.encoder->input[0], pipuck_mmal->internal.pool_in);
+  mmal_port_pool_destroy(pipuck_mmal->internal.encoder->output[0], pipuck_mmal->internal.pool_out);
+  vcos_semaphore_delete(&(pipuck_mmal->internal.semaphore));
+}
+
+int pipuck_mmal_convert(pipuck_mmal_t *pipuck_mmal) {
+  MMAL_BOOL_T eos = MMAL_FALSE;
+  int sent_bytes = 0;
+  MMAL_BUFFER_HEADER_T *buffer;
+  MMAL_STATUS_T status;
+
+  pipuck_image_t *output_image = &(pipuck_mmal->output);
+  pipuck_image_t *input_image = &(pipuck_mmal->input);
+  pipuck_mmal_internal_t *context = &(pipuck_mmal->internal);
+
+  output_image->size = 0;
+
+  while (eos == MMAL_FALSE) {
+    VCOS_STATUS_T vcos_status = vcos_semaphore_wait_timeout(&(context->semaphore), 2000);
+    assert(vcos_status == VCOS_SUCCESS);
+
+    // Input frames
+    if ((buffer = mmal_queue_get(context->pool_in->queue)) != NULL) {
+      buffer->length = MAX(0, MIN(buffer->alloc_size, input_image->size - sent_bytes));
+      buffer->offset = 0;
+      buffer->pts = MMAL_TIME_UNKNOWN;
+      buffer->dts = MMAL_TIME_UNKNOWN;
+      buffer->data = input_image->data + sent_bytes;
+      sent_bytes += buffer->length;
+      status = mmal_port_send_buffer(context->encoder->input[0], buffer);
+      assert(status == MMAL_SUCCESS);
+    }
+
+    // Output frames
+    while ((buffer = mmal_queue_get(context->queue)) != NULL) {
+      eos = buffer->flags & MMAL_BUFFER_HEADER_FLAG_FRAME_END;
+      memcpy(output_image->data + output_image->size, buffer->data, buffer->length);
+      output_image->size += buffer->length;
+      mmal_buffer_header_release(buffer);
+    }
+
+    // Send empty buffers to the output port of the decoder
+    while ((buffer = mmal_queue_get(context->pool_out->queue)) != NULL) {
+      status = mmal_port_send_buffer(context->encoder->output[0], buffer);
+      assert(status == MMAL_SUCCESS);
+    }
+  }
+
+  return 0;
```

Comment:
Is it really necessary to return an int since it returns always ``0``?
And the return value is never used.

Reply:
You are right! No need, it supposed to be an error code, but no need since there are asserts.